### PR TITLE
Optimize time complexity of broadcasts via better data structure for streams

### PR DESCRIPTION
**Pull request by cmyui**

### CODE REVIEW

File: events/loginEvent.py [hunk 1]
```diff
@@ -496,7 +496,10 @@ async def handle(web_handler: AsyncRequestHandler) -> tuple[str, bytes]:  # toke
 
         # Send to everyone our userpanel if we are not restricted or tournament
         if not osuToken.is_restricted(userToken["privileges"]):
-            await streamList.broadcast("main", await serverPackets.userPanel(userID))
+            await stream_messages.broadcast_data(
+                "main",
+                await serverPackets.userPanel(userID),
```
Comment (tsunyoku) [inline events/loginEvent.py:501]: unrelated to this PR but seeing a packet which is obviously doing more work than construction (the await) makes me scared lol

Comment (cmyui) [reply events/loginEvent.py:501]: lol yep, i knew about these from a while back. to be fixed at some point.

File: helpers/chatHelper.py [hunk 2]
```diff
@@ -427,16 +420,9 @@ async def _handle_public_message(
...
-        for token in await osuToken.get_tokens()
-        if (
-            # Never send messages to any chatbot sessions
-            token["user_id"] != CHATBOT_USER_ID
-            # Never send our messages to our own session
-            and token["token_id"] != sender_token["token_id"]
-        )
-    ]
+    # TODO: token["user_id"] != CHATBOT_USER_ID
+    # TODO: and token["token_id"] != sender_token["token_id"]
```
Comment (cmyui) [inline helpers/chatHelper.py]: Verify before merge

File: objects/osuToken.py [hunk 3]
```diff
@@ -440,10 +442,19 @@ async def get_streams(token_id: str) -> set[str]:
 
 async def add_stream(token_id: str, stream_name: str) -> None:
     await glob.redis.sadd(f"{make_key(token_id)}:streams", stream_name)
+    await glob.redis.hset(
+        f"{make_key(token_id)}:stream_offsets",
+        stream_messages.make_key(stream_name),
+        "0-0",  # TODO: get the current offset of the stream
```
Comment (cmyui) [inline objects/osuToken.py]: Needs fix before merge

File: objects/stream_messages.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    token_id: str,
+    stream_offsets: dict[str, str],
+) -> None:
+    await glob.redis.hmset(
+        f"bancho:tokens:{token_id}:stream_offsets",
+        stream_offsets,  # type: ignore[arg-type]
+    )
+
+
+async def read_all_pending_data(token_id: str) -> bytes:
```
Comment (cmyui) [inline objects/stream_messages.py:66]: image

File: objects/stream_messages.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    return f"bancho:streams:{stream_name}:messages"
+
+
+class StreamMessage(TypedDict):
+    stream_key: str
+    packet_data: bytes
+    excluded_token_ids: str
+
+
+async def broadcast_data(
```
Comment (cmyui) [inline objects/stream_messages.py:20]: image

File: helpers/chatHelper.py (hunk 2 above)
Comment (cmyui) [inline helpers/chatHelper.py]:
```suggestion
```

File: objects/stream_messages.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    stream_name: str,
+    data: bytes,
+    *,
+    excluded_token_ids: list[str] | None = None,
+) -> None:
+    """Send some data to all clients connected to this stream, with optional exclusions"""
+    if excluded_token_ids is None:
+        excluded_token_ids = []
+
+    if not await streamList.stream_exists(stream_name):
```
Comment (cmyui) [inline objects/stream_messages.py:31]:
```suggestion
    # TODO: potentially remove this check? it's ~55% of the function's wall time
    if not await streamList.stream_exists(stream_name):
```

**Pull request merged**